### PR TITLE
Support 20 bytes address

### PR DESCRIPTION
**Pull request by lxfind**
The current address is hard-coded to be 16 bytes. This works well for Diem, but could be insufficient for other systems. A common requirement is 20 bytes.
This PR adds a feature that gates the size of the address to be either 16 or 20 bytes, and change everywhere this is used to be using the right length. If we want to support arbitrary configurable size, we would need to add a build.rs to generate it. Most of the changes here are still needed.
There are a few more things to polish/warnings to fix. 
Sending out early to get feedback and see if we think this is something reasonable to do.

### CODE REVIEW

**Comment by bmwill** [comment]
While you can use features this way it's generally discouraged because you'd need a feature flag for each length and then the features are no longer additive. 

I wonder if a more scalable solution would be to use const generics to determine the size of the address.

**Comment by lxfind** [comment]
> I wonder if a more scalable solution would be to use const generics to determine the size of the address.

Would love to make this more generic. Thought about const generics, but couldn't see a way to do this without changing everything (all uses of AccountAddress and classes that have it). Would love to hear ideas on how to do this!
I could use a build.rs to generate the length definition based on a config, if that works for the team.

**Comment by wrwg** [comment]
> > I wonder if a more scalable solution would be to use const generics to determine the size of the address.
> 
> Would love to make this more generic. Thought about const generics, but couldn't see a way to do this without changing everything (all uses of AccountAddress and classes that have it). Would love to hear ideas on how to do this! I could use a build.rs to generate the length definition based on a config, if that works for the team.

I think any technique, let it be generated build.rs or generic const or like done here, will lead to that there are two exclusive flavors of the Move binaries. The only way out of this seems to me dynamically sized addresses, which has performance and other far reaching implications, so I guess I'm good with this one here.

**Comment by tnowacki** [comment]
> I could use a build.rs to generate the length definition based on a config, if that works for the team.

Could you show more what that would look like? Even a mock up of some sorts is fine for now.

Just not entirely sure what goes in the `build.rs`

**Comment by lxfind** [comment]
Turns out that I don't have to use build.rs (I was initially thinking about having build.rs generating a rs file that contains a definition of LENGTH based on an environment variable).
I discovered a crate called "const_env", which allows you to declare a const integer from an environment variable.
Updated the change. Now we could specify the address length through environment variable ADDRESS_LENGTH. It's default to 16.

**Comment by wrwg** [comment]
> Turns out that I don't have to use build.rs (I was initially thinking about having build.rs generating a rs file that contains a definition of LENGTH based on an environment variable). I discovered a crate called "const_env", which allows you to declare a const integer from an environment variable. Updated the change. Now we could specify the address length through environment variable ADDRESS_LENGTH. It's default to 16.

I think this just makes it worse compared to the exclusive cfg feature you have used before. const_env evaluates environment variables at compile time. Thus compilation of the Move code now depends magically on some environment variable, which is less transparent than the cargo feature, but comes exactly with the same disadvantages which are [listed for exclusive features here](https://doc.rust-lang.org/cargo/reference/features.html#mutually-exclusive-features).

With the feature, one can at least specify in the toml on which flavor of Move one wants to depend. With the environment variable, it is even not possible to request an address flavor via a dependency.  

While exclusive features are discouraged in general,  there are cases (like this here) where they are needed. Lets use a feature instead of some obscure dependency on an env variable at compile time. As I said, the only way to do this better than with features is via dynamic (at runtime) configurable address length, but this is to big of a refactoring.

**Comment by tnowacki** [comment]
I'd still like to better understand how the build.rs solution would work, but I agree with @wrwg that the env based approach is much much worse. 

build.rs aside. Could we at least make the cfg feature based solution cleaner?
I think in the previous iteration, @lxfind, you had conditionally declared the constant depending on the feature. While functionally equivalent, could we do something like `const C: u64 = if cfg!(feature = "my_cool_addr_length_feature") { 0 } else { 1 };` instead?

**Comment by lxfind** [comment]
> I'd still like to better understand how the build.rs solution would work, but I agree with @wrwg that the env based approach is much much worse.
> 
> build.rs aside. Could we at least make the cfg feature based solution cleaner? I think in the previous iteration, @lxfind, you had conditionally declared the constant depending on the feature. While functionally equivalent, could we do something like `const C: u64 = if cfg!(feature = "my_cool_addr_length_feature") { 0 } else { 1 };` instead?

Thanks for the feedback! @wrwg @tnowacki 
The approach with build.rs will only be worse in this regard (I was planning to generate a rust file that contains the definition of LENGTH based on environment variables and include it in account_address. So I won't go that route.

The closest I can do with if/else would be something like this:
```
pub const LENGTH: usize = match () {
        #[cfg(feature = "address20")]
        () => 20,
        #[cfg(not(feature = "address20"))]
        () => 16,
    };
```
which isn't much better than
```
#[cfg(feature = "address20")]
pub const LENGTH: usize = 20;
#[cfg(not(feature = "address20"))]
pub const LENGTH: usize = 16;
```

**Comment by tnowacki** [comment]
> > I'd still like to better understand how the build.rs solution would work, but I agree with @wrwg that the env based approach is much much worse.
> > build.rs aside. Could we at least make the cfg feature based solution cleaner? I think in the previous iteration, @lxfind, you had conditionally declared the constant depending on the feature. While functionally equivalent, could we do something like `const C: u64 = if cfg!(feature = "my_cool_addr_length_feature") { 0 } else { 1 };` instead?
> 
> Thanks for the feedback! @wrwg @tnowacki The approach with build.rs will only be worse in this regard (I was planning to generate a rust file that contains the definition of LENGTH based on environment variables and include it in account_address. So I won't go that route.
> 
> The closest I can do with if/else would be something like this:
> 
> ```
> pub const LENGTH: usize = match () {
>         #[cfg(feature = "address20")]
>         () => 20,
>         #[cfg(not(feature = "address20"))]
>         () => 16,
>     };
> ```
> 
> which isn't much better than
> 
> ```
> #[cfg(feature = "address20")]
> pub const LENGTH: usize = 20;
> #[cfg(not(feature = "address20"))]
> pub const LENGTH: usize = 16;
> ```

Sorry, I don't understand why you can't do 
```
pub const LENGTH: usize = if cfg!(feature = "address20") { 20 } else { 16 };
``` 
Or 
```
pub const LENGTH: usize = if cfg!(feature = "address20") {
    20
} else if cfg!(feature = "address16") {
    16
} else {
    panic!("address feature not specified")
};
```

**Comment by lxfind** [comment]
> Sorry, I don't understand why you can't do
> 
> ```
> pub const LENGTH: usize = if cfg!(feature = "address20") { 20 } else { 16 };
> ```
> 
> Or
> 
> ```
> pub const LENGTH: usize = if cfg!(feature = "address20") {
>     20
> } else if cfg!(feature = "address16") {
>     16
> } else {
>     panic!("address feature not specified")
> };
> ```

Ha didn't know cfg! could be used this way! TIL. Will use it. Thanks!

**Comment by wrwg** [comment]
> > Sorry, I don't understand why you can't do
> > ```
> > pub const LENGTH: usize = if cfg!(feature = "address20") { 20 } else { 16 };
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Or
> > ```
> > pub const LENGTH: usize = if cfg!(feature = "address20") {
> >     20
> > } else if cfg!(feature = "address16") {
> >     16
> > } else {
> >     panic!("address feature not specified")
> > };
> > ```
> 
> Ha didn't know cfg! could be used this way! TIL. Will use it. Thanks!

Instead of panic!, you may want to use `compile_error!("msg")`

**Comment by libra-action** [comment]
**LSR dependency change summary**:
```
target packages:
  M diem-types 0.0.3 (workspace, path 'types')
    * added features: address20
    * (unchanged features: default)
```

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  M diem-types 0.0.3 (workspace, path 'types')
    * added features: address20
    * (unchanged features: default)
  M num-bigint 0.4.0 (direct third-party, crates.io)
    * status changed from transitive third-party
    * added features: default
    * (unchanged features: std)
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  M diem-types 0.0.3 (workspace, path 'types')
    * added features: address20
    * (unchanged features: default)
  M num-bigint 0.4.0 (direct third-party, crates.io)
    * status changed from transitive third-party
    * added features: default
    * (unchanged features: std)
```

**Comment by libra-action** [comment]
**LSR dependency change summary**:
```
target packages:
  M diem-types 0.0.3 (workspace, path 'types')
    * added features: address20
    * (unchanged features: default)
```

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  M diem-types 0.0.3 (workspace, path 'types')
    * added features: address20
    * (unchanged features: default)
  M num-bigint 0.4.0 (direct third-party, crates.io)
    * status changed from transitive third-party
    * added features: default
    * (unchanged features: std)
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  M diem-types 0.0.3 (workspace, path 'types')
    * added features: address20
    * (unchanged features: default)
  M num-bigint 0.4.0 (direct third-party, crates.io)
    * status changed from transitive third-party
    * added features: default
    * (unchanged features: std)
```

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  M num-bigint 0.4.0 (direct third-party, crates.io)
    * status changed from transitive third-party
    * added features: default
    * (unchanged features: std)
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  M num-bigint 0.4.0 (direct third-party, crates.io)
    * status changed from transitive third-party
    * added features: default
    * (unchanged features: std)
```

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  M num-bigint 0.4.0 (direct third-party, crates.io)
    * status changed from transitive third-party
    * added features: default
    * (unchanged features: std)
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  M num-bigint 0.4.0 (direct third-party, crates.io)
    * status changed from transitive third-party
    * added features: default
    * (unchanged features: std)
```

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  M num-bigint 0.4.0 (direct third-party, crates.io)
    * status changed from transitive third-party
    * added features: default
    * (unchanged features: std)
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  M num-bigint 0.4.0 (direct third-party, crates.io)
    * status changed from transitive third-party
    * added features: default
    * (unchanged features: std)
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  M num-bigint 0.4.0 (direct third-party, crates.io)
    * status changed from transitive third-party
    * added features: default
    * (unchanged features: std)
```

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  M num-bigint 0.4.0 (direct third-party, crates.io)
    * status changed from transitive third-party
    * added features: default
    * (unchanged features: std)
```

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  M num-bigint 0.4.0 (direct third-party, crates.io)
    * status changed from transitive third-party
    * added features: default
    * (unchanged features: std)
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  M num-bigint 0.4.0 (direct third-party, crates.io)
    * status changed from transitive third-party
    * added features: default
    * (unchanged features: std)
```

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  M num-bigint 0.4.0 (direct third-party, crates.io)
    * status changed from transitive third-party
    * added features: default
    * (unchanged features: std)
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  M num-bigint 0.4.0 (direct third-party, crates.io)
    * status changed from transitive third-party
    * added features: default
    * (unchanged features: std)
```

**Comment by lxfind** [comment]
All tests finally passing now. Ready for review. I initially wanted to make it such that it can pass all tests when `address20` is enabled. But realized that it would be likely impractical. The move release are precompiled modules that use 16 bytes, and they cannot work with 20 bytes. So my current goal is just to make sure that move builds with `address20`, and we will test everything separately when depend on it.

File: language/move-core/types/src/account_address.rs [hunk 1]
```diff
@@ -17,11 +17,21 @@ impl AccountAddress {
     }
 
     /// The number of bytes in an address.
-    pub const LENGTH: usize = 16;
+    /// Default to 16 bytes, can be set to 20 bytes with address20 feature.
+    pub const LENGTH: usize = if cfg!(feature = "address20") { 20 } else { 16 };
```
Comment (tnowacki) [inline language/move-core/types/src/account_address.rs:21]: Everything in this PR looks good, not sure if @wrwg or @bmwill have any final thoughts on this direction?

**Review by sblackshear — APPROVED** [review]
Approving this in the absence of objections!

**Pull request closed without merging**